File: solutions/cpp/search_frequent_items.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::ios;
using std::istringstream;
using std::random_device;
using std::string;
using std::uniform_int_distribution;
using std::unordered_map;
using std::vector;

string RandString(int len) {
  string ret;
  default_random_engine gen((random_device())());
  while (len--) {
    uniform_int_distribution<int> dis('a', 'z');
    ret += dis(gen);
  }
  return ret;
}

// @include
// Finds the candidates which may occur > n / k times.
vector<string> SearchFrequentItems(int k, istringstream* stream) {
  string buf;
  unordered_map<string, int> table;
  int n = 0;  // Count the number of strings.

  while (*stream >> buf) {
    ++table[buf], ++n;
    // Detecting k items in table, at least one of them must have exactly
    // one in it. We will discard those k items by one for each.
    if (table.size() == k) {
      auto it = table.begin();
      while (it != table.end()) {
        if (--(it->second) == 0) {
          table.erase(it++);
        } else {
          ++it;
        }
      }
    }
  }

  // Resets table for the following counting.
  for (auto& it : table) {
    it.second = 0;
  }

  // Resets the stream and read it again.
  stream->clear();
  stream->seekg(0, ios::beg);
  // Counts the occurrence of each candidate word.
  while (*stream >> buf) {
    auto it = table.find(buf);
    if (it != table.end()) {
      ++it->second;
    }
  }

  // Selects the word which occurs > n / k times.
  vector<string> ret;
  for (const auto& it : table) {
    if (it.second > static_cast<double>(n) / k) {
      ret.emplace_back(it.first);
    }
  }
  return ret;
}
// @exclude

void CheckAns(vector<string>* stream, int k, vector<string>* items) {
  sort(items->begin(), items->end());
  sort(stream->begin(), stream->end());
  int count = 1, idx = 0;
  for (int i = 1; i < stream->size(); ++i) {
    if ((*stream)[i].compare((*stream)[i - 1])) {
      if (count > static_cast<double>(stream->size()) / k) {
        assert(idx < items->size());
        assert((*stream)[i - 1].compare((*items)[idx++]) == 0);
      }
      count = 1;
    } else {
      ++count;
    }
  }
  if (count > static_cast<double>(stream->size()) / k) {
    assert((*stream).back().compare((*items)[idx++]) == 0);
  }
  assert(idx == items->size());
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    cout << times << endl;
    vector<string> stream;
    int n, k;
    if (argc == 2) {
      n = atoi(argv[1]);
      uniform_int_distribution<int> k_dis(1, n);
      k = k_dis(gen);
    } else if (argc == 3) {
      n = atoi(argv[1]);
      k = atoi(argv[2]);
    } else {
      uniform_int_distribution<int> n_dis(0, 99999);
      n = n_dis(gen);
      uniform_int_distribution<int> k_dis(1, n);
      k = k_dis(gen);
    }
    for (int i = 0; i < n; ++i) {
      uniform_int_distribution<int> dis(1, 5);
      stream.emplace_back(RandString(dis(gen)));
    }
    string s;
    for (int i = 0; i < stream.size(); ++i) {
      s += stream[i];
      s += ' ';
    }
    istringstream stream(s);
    vector<string> items = SearchFrequentItems(k, &stream);
    CheckAns(&stream, k, &items);
  }
  return 0;
}

File: solutions/java/SearchFrequentItems.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class SearchFrequentItems {
  public static String randString(int len) {
    StringBuilder sb = new StringBuilder();
    Random gen = new Random();
    while (len-- != 0) {
      sb.append((char)(gen.nextInt('z' + 1 - 'a') + 'a'));
    }
    return sb.toString();
  }

  // @include
  public static List<String> searchFrequentTtems(Iterable<String> stream,
                                                 int k) {
    // Finds the candidates which may occur > n / k times.
    String buf = "";
    Map<String, Integer> table = new HashMap<>();
    int n = 0; // Counts the number of strings.

    Iterator<String> sequence = stream.iterator();
    while (sequence.hasNext()) {
      buf = sequence.next();
      table.put(buf, table.containsKey(buf) ? table.get(buf) + 1 : 1);
      ++n;
      // Detecting k items in table, at least one of them must have exactly one
      // in it. We will discard those k items by one for each.
      if (table.size() == k) {
        List<String> delKeys = new ArrayList<>();
        for (Map.Entry<String, Integer> entry : table.entrySet()) {
          if (entry.getValue() - 1 == 0) {
            delKeys.add(entry.getKey());
          } else {
            table.put(entry.getKey(), entry.getValue() - 1);
          }
        }
        for (String s : delKeys) {
          table.remove(s);
        }
      }
    }

    // Resets table for the following counting.
    for (String it : table.keySet()) {
      table.put(it, 0);
    }

    // Counts the occurrence of each candidate word.
    sequence = stream.iterator();
    while (sequence.hasNext()) {
      buf = sequence.next();
      Integer it = table.get(buf);
      if (it != null) {
        table.put(buf, it + 1);
      }
    }

    // Selects the word which occurs > n / k times.
    List<String> ret = new ArrayList<>();
    for (Map.Entry<String, Integer> it : table.entrySet()) {
      if (n * 1.0 / k < (double)it.getValue()) {
        ret.add(it.getKey());
      }
    }

    return ret;
  }
  // @exclude

  public static void checkAns(List<String> stream, int k, List<String> items) {
    Collections.sort(stream);
    Collections.sort(items);

    int count = 1, idx = 0;
    for (int i = 1; i < stream.size(); ++i) {
      if (stream.get(i).compareTo(stream.get(i - 1)) != 0) {
        if ((double)count > 1.0 * stream.size() / k) {
          assert(idx < items.size());
          assert(stream.get(i - 1).compareTo(items.get(idx++)) == 0);
        }
        count = 1;
      } else {
        ++count;
      }
    }
    if ((double)count > 1.0 * stream.size() / k) {
      assert(stream.get(stream.size() - 1).compareTo(items.get(idx++)) == 0);
    }
    assert(idx == items.size());
  }

  public static void main(String[] args) {
    Random gen = new Random();
    for (int times = 0; times < 1000; ++times) {
      System.out.println(times);
      List<String> stream = new ArrayList<>();
      int n, k;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
        k = gen.nextInt(n) + 1;
      } else if (args.length == 2) {
        n = Integer.parseInt(args[0]);
        k = Integer.parseInt(args[1]);
      } else {
        n = gen.nextInt(99999);
        k = gen.nextInt(n) + 1;
      }
      for (int i = 0; i < n; ++i) {
        stream.add(randString(gen.nextInt(5) + 1));
      }
      List<String> items = searchFrequentTtems(stream, k);
      checkAns(stream, k, items);
    }
  }
}
